papers/plutus-platform/authoring.tex

```diff
@@ -0,0 +1,307 @@
+
+How do we deal with the extra features of GHC Core?
+First, we split the problem in half, by defining an intermediate language, \gls{plutus-ir} (see \cref{sec:plutus-ir}), which is much closer to GHC Core.
+Most of the theoretical complexity is therefore moved to the \gls{plutus-ir} compiler.
+
+The remaining work of the \gls{plutus-tx} compiler is then to \emph{lower} the GHC Core terms and types into their corresponding \gls{plutus-ir} variants, emitting errors as appropriate if we encounter features we do not support.
+
+\subsubsection{Supporting Haskell's features}
+
+As alluded to in the previous section, we do not support the entirety of Haskell.
+Thanks to the design of GHC, we get a great deal for free, as we compile programs after they have been converted to GHC Core, which means that most of the complex source-level features of Haskell have already been desugared into a smaller set of simpler features.
+
+While we support most ``standard'' Haskell, there are quite a few things we do not support. A non-exhaustive list of features that we do not support is:
+\begin{itemize}
+\item Not implemented yet
+  \begin{itemize}
+  \item Mutually recursive datatypes (should be done by release)
+  \end{itemize}
+\item Incompatible with the design of Plutus Core
+  \begin{itemize}
+  \item \code{PolyKinds}, \code{DataKinds}, anything that moves towards "Dependent Haskell"
+  \end{itemize}
+\item Technically difficult
+  \begin{itemize}
+  \item Literal patterns
+  \end{itemize}
+\item Requires access to function definitions (might be fixed with some GHC work)
+  \begin{itemize}
+  \item Function usage without \code{INLINEABLE} or \code{-fexpose-all-unfoldings}
+  \item Typeclass dictionaries
+  \end{itemize}
+\item Use of coercions required
+  \begin{itemize}
+  \item GADTs
+  \item \code{Data.Coerce}
+  \item \code{DerivingVia}, \code{GeneralizedNewtypeDeriving}, etc.
+  \end{itemize}
+\item Assumes "normal" codegen
+  \begin{itemize}
+  \item FFI
+  \item Numeric types other than integers
+  \item Unlifted/\code{MagicHash} types
+  \item Machine words, C strings, etc.
+  \end{itemize}
+\end{itemize}
+
+\todompj{Perhaps expand into multiple sections with brief explanations?}
+
+\subsubsection{Strictness}
+\label{sec:plutus-tx-strictness}
+
+Haskell is a lazy language and \gls{plutus-core} is a strict language.
+How can we compile a lazy language into a strict language efficiently?
+
+The answer is that we handle this partially.
+We generally compile Haskell as though it were strict, but the key exception is for non-value let-bindings.
+That is, if we see a let-binding whose right-hand-side is not a value (i.e. may evaluate further), then we compile it as a non-strict let-binding (see \cref{sec:pir-non-strict}).
+
+Overall, we have no proof that our approach is sound.
+
+\todompj{This sounds terrible. I mean, it \emph{is} pretty terrible, but maybe we can make it sound better.}
```

Comment:
If not, then maybe just delete the entire line (l. 184)

Reply:
I kind of feel like it's important to be honest about this, I think it's a serious issue.